src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -1328,5 +1351,18 @@ def _do_upgrade(self):
                 'who': name_to_config_section(daemon_type),
             })
 
+        # Upgrade is complete. Unset mon flags
+        if self.upgrade_state.mon_flags:
```

Comment:
this handles the successfully completed upgrade case, but I think we additionally need to handle the case where users manually stop the upgrade before it's completed.

We should also consider the staggered upgrade scenario. If a user does a staggered upgrade, say for just the mgr daemons, would we want to unset the flags when that completes and then set them again when they trigger an additional staggered upgrade for the next daemon type/service/host etc.? It doesn't feel totally correct, but it might be difficult to do it any other way (as in, "knowing" there will be additional staggered upgrades triggered and not unsetting the flags) so we might have to handle it like that.

Reply:
I've handled full upgrades, staggered upgrades, manual upgrade interruptions and upgrade failures. Could you please review this again and see if it suffices?